includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,384 @@
+              '@endpoint' => $endpoint, 
+              '@expectation' => $expectation,
+              '@response' => $response,
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+          watchdog($messages);
+        }
+      } else {
+        $messages = array(
+          'message' => t("Expected JSON key from @endpoint but could not decode. @response"),
+          'message_sub' => array(
+            '@endpoint' => $endpoint,
+            '@response' => $response,
+          ),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+        watchdog($messages);
+      }
+    } else {
+      $messages = array(
+        'message' => t("Expected one of @expected; got a @response_code"),
+        'message_sub' => array(
+          '@expected' => implode(',', array_keys($expect)),
+          '@response_code' => $response_code,
+        ), 
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+      watchdog($messages);
+    }
+  }
+  return;
+}
+
+/**
+ * Uploads file to Panopto using AWS S3 Uploader
+ * @param string $upload_target The URL of upload target
+ * @param string $file_path The URL which is the target of the request
+ * @return void
+ */
+function islandora_video_streaming_panopto_upload_file($upload_target, $file_path) {
+  $element = explode("/", $upload_target);
+  $prefix = array_pop($element);
+  $service_endpoint = implode("/", $element);
+  $bucket = array_pop($element);
+  $file_name = basename($file_path);
+  $object_key = $prefix."/".$file_name;
+  
+  $s3Params = array(
+      'endpoint' => $service_endpoint,
+      'region'  => 'us-east-1',
+      'version' => '2006-03-01',
+      'credentials' => [
+        'key'    => 'dummy',
+        'secret' => 'dummy'
+      ]
+  );
+  // Create an S3Client
+  $s3Client = new S3Client($s3Params);
+
+  $source = fopen($file_path, 'rb');
+  $uploader = new ObjectUploader(
+    $s3Client,
+    $bucket,
+    $object_key,
+    $source
+  );
+
+  do {
+    try {
+      $result = $uploader->upload();
+    } catch (MultipartUploadException $e) {
+      rewind($source);
+      $uploader = new MultipartUploader($s3Client, $source, [
+        'state' => $e->getState(),
+      ]);
+    }
+  } while (!isset($result));
+}
+
+/**
+ * Creates manifest file, in temporary storage, needed for Panopto Video Upload
+ * @param string $file_path The filepath of the Islandora object
+ * @return string The filename or filepath of the newly created manifest file
+ */
+function islandora_video_streaming_panopto_create_manifest_for_video($file_path = null) {
+  $file_name = basename($file_path);
+  $manifest_template = drupal_get_path('module','islandora_video_streaming') .'/includes/' . 'upload_manifest_template.xml';
```

Comment:
The slashes in a file path are always better as the constant `DIRECTORY_SEPARATOR`, so that the code is compatible across Linux and Windows paths.